service/worker/deployment/deployment_client.go

```diff
@@ -369,51 +338,396 @@ func (d *DeploymentClientImpl) ListDeployments(ctx context.Context, namespaceEnt
 
 }
 
-// GenerateStartWorkflowPayload generates start workflow execution payload
-func (d *DeploymentClientImpl) generateStartWorkflowPayload(namespaceEntry *namespace.Namespace, deployment *deploymentpb.Deployment) (*commonpb.Payloads, error) {
-	workflowArgs := &deploymentspb.DeploymentWorkflowArgs{
+func (d *DeploymentClientImpl) SetCurrentDeployment(
+	ctx context.Context,
+	namespaceEntry *namespace.Namespace,
+	deployment *deploymentpb.Deployment,
+	updateMetadata *deploymentpb.UpdateDeploymentMetadata,
+	identity string,
+	requestID string,
+) (*deploymentpb.DeploymentInfo, *deploymentpb.DeploymentInfo, error) {
+	updatePayload, err := sdk.PreferProtoDataConverter.ToPayloads(&deploymentspb.SetCurrentDeploymentArgs{
+		Identity:       identity,
+		BuildId:        deployment.BuildId,
+		UpdateMetadata: updateMetadata,
+		RequestId:      requestID,
+	})
+	outcome, err := d.updateWithStartDeploymentSeries(
+		ctx,
+		namespaceEntry,
+		deployment.SeriesName,
+		&updatepb.Request{
+			Input: &updatepb.Input{Name: SetCurrentDeployment, Args: updatePayload},
+			Meta:  &updatepb.Meta{UpdateId: requestID, Identity: identity},
+		},
+		identity,
+		requestID,
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if failure := outcome.GetFailure(); failure.GetApplicationFailureInfo().GetType() == errNoChangeType {
+		return nil, nil, serviceerror.NewAlreadyExist(fmt.Sprintf("Build ID %q is already current for %q",
+			deployment.BuildId, deployment.SeriesName))
+	} else if failure != nil {
+		// TODO: is there an easy way to recover the original type here?
+		return nil, nil, serviceerror.NewInternal(failure.Message)
```

Comment:
What kinds of errors these would be? should we log it?

Reply:
Whatever errors handleSetCurrent in the series wf can return, which is quite a bunch of different ones along the way, since it does an activity, two more wf updates, more activities, etc. Mostly stuff that shouldn't ever happen like payload conversion, namespace lookups, unexpected proto layout in update result. But also errors updating matching user data.